### ISSUE TITLE
Domoticz crashes after OS time change

### ISSUE THREAD

**Issue by sander1988** [domoticz/domoticz#4734]
Hardware: RPI 4
Version: 2020.2
Build Hash: b63341bc0
Compile Date: 2020-04-26 13:47:55
dzVents Version: 3.0.2
Python Version: 3.7.3 (default, Jul 25 2020, 13:03:44) [GCC 8.3.0]

---

This morning my Domoticz crashed (due to the build-in watchdog) after I changed my OS time:
```
2021-03-17 08:32:19.370  Error: mainworker seems to have ended or hung unexpectedly (last update 907.000000 seconds ago)
2021-03-17 08:32:19.370  Error: Domoticz(940) is exiting due to watchdog triggered...
2021-03-17 08:32:19.402  Error: Failed to start gdb, will use backtrace() for printing stack frame
```

I checked the source code (https://github.com/domoticz/domoticz/blob/master/main/SignalHandler.cpp#L496) and I see the current timestamp is used here. Best practices say you should use the CPU cycles count here (C function clock() for example).

**Comment by gizmocuz** [comment]
Well to be honest, I never likes this function (calling raise/abort)
The issue here is that the timeout is checked against the value 300, when the clock skips 3600 seconds (an hour), this happens
It can also be solved differently... add a counter here, when there is a timeout, count+1, next time when counter !=0, perform the action.

I am not sure if 'clock()' would work, we also want to check if the 'mainworker' is still running (or being stopped by another process like a plugin for example)

Open for suggestions/PR

Looking at your timestamp, I think something else went wront... your system time does not change at 08:32am ...

**Comment by sander1988** [comment]
> Looking at your timestamp, I think something else went wront... your system time does not change at 08:32am ...

I shouldn't have copied only the 3 log lines containing the crash, this is the full log:
```
2021-03-17 08:17:12.502  Status: Build Hash: b63341bc0, Date: 2020-04-26 13:47:55
2021-03-17 08:17:12.503  Status: Startup Path: /opt/domoticz/
2021-03-17 08:17:12.644  Sunrise: 06:48:00 SunSet: 18:47:00
2021-03-17 08:17:12.644  Day length: 11:59:00 Sun at south: 12:47:00
2021-03-17 08:17:12.644  Civil twilight start: 06:14:00 Civil twilight end: 19:20:00
2021-03-17 08:17:12.644  Nautical twilight start: 05:35:00 Nautical twilight end: 20:00:00
2021-03-17 08:17:12.644  Astronomical twilight start: 04:54:00 Astronomical twilight end: 20:41:00
2021-03-17 08:17:13.004  Status: PluginSystem: Started, Python version '3.7.3'.
2021-03-17 08:17:13.025  Active notification Subsystems: pushover (1/13)
2021-03-17 08:17:13.063  Status: WebServer(HTTP) started on address: 0.0.0.0 with port 8080
2021-03-17 08:17:13.085  Starting shared server on: 0.0.0.0:6144
2021-03-17 08:17:13.085  Status: TCPServer: shared server started...
2021-03-17 08:17:13.086  Status: RxQueue: queue worker started...
2021-03-17 08:17:15.089  Status: (TV Woonkamer) Started.
2021-03-17 08:17:15.089  Status: (PC Sander) Started.
2021-03-17 08:17:15.090  Status: (NAS) Started.
2021-03-17 08:17:15.090  Status: (Server) Started.
2021-03-17 08:17:15.090  Status: (HCTasmota) Started.
2021-03-17 08:17:15.090  Status: (HCNode) Started.
2021-03-17 08:17:15.090  Status: (HCPresence) Started.
2021-03-17 08:17:15.090  Status: (Roomba) Started.
2021-03-17 08:17:15.090  Status: (UPS) Started.
2021-03-17 08:17:15.090  Status: (Grasmaaier) Started.
2021-03-17 08:17:15.090  Status: (Shield Woonkamer) Started.
2021-03-17 08:17:15.090  Status: NotificationSystem: thread started...
2021-03-17 08:17:15.092  Status: Darksky: Started...
2021-03-17 08:17:15.094  Status: EventSystem: reset all events...
2021-03-17 08:17:15.147  Status: dzVents: Write file: /var/lib/domoticz/scripts/dzVents/generated_scripts/FIRST.lua
-- private scripts removed from log --
2021-03-17 08:17:15.154  Status: dzVents: Write file: /var/lib/domoticz/scripts/dzVents/generated_scripts/LAST.lua
2021-03-17 08:17:15.154  Status: EventSystem: reset all device statuses...
2021-03-17 08:17:15.248  Status: PluginSystem: Entering work loop.
2021-03-17 08:17:15.296  Status: RFLink: Using serial port: /dev/ttyACM0
2021-03-17 08:32:18.949  Status: MQTT: Connecting to 127.0.0.1:1883
2021-03-17 08:32:19.067  Status: RFLink: Controller Initialized!...
2021-03-17 08:32:19.143  Status: RFLink Detected, Version: 1.1 Revision: 48 Build: 4
2021-03-17 08:32:19.208  Status: MQTT: connected to: 127.0.0.1:1883
2021-03-17 08:32:19.370  Error: mainworker seems to have ended or hung unexpectedly (last update 907.000000 seconds ago)
2021-03-17 08:32:19.370  Error: Domoticz(940) is exiting due to watchdog triggered...
2021-03-17 08:32:19.402  Error: Failed to start gdb, will use backtrace() for printing stack frame

2021-03-17 08:32:19.410  Status: MQTT: Subscribed
2021-03-17 08:32:19.936  Status: Closing application!...
2021-03-17 08:32:19.936  Status: Stopping worker...
2021-03-17 08:32:20.222  Status: Incoming connection from: 127.0.0.1
2021-03-17 08:32:20.250  Status: dzVents: Info: Handling system event for: "stop"
2021-03-17 08:32:20.250  Status: dzVents: Info: Garden-Watering: ------ Start internal script: Garden-Watering: Domoticz event: "stop"
2021-03-17 08:32:20.321  Status: (TV Woonkamer) Entering work loop.
2021-03-17 08:32:20.321  Status: (TV Woonkamer) Initialized version 3.0, author 'Sander Peterse'
2021-03-17 08:32:20.329  Status: dzVents: Info: Garden-Watering: ------ Finished Garden-Watering
2021-03-17 08:32:20.336  Status: RxQueue: queue worker stopped...
2021-03-17 08:32:20.360  Status: Python EventSystem: Initalizing event module.
2021-03-17 08:32:20.377  Status: EventSystem: Queue thread started...
2021-03-17 08:32:20.387  Status: EventSystem: Started
...
```

And note that I'm 100% sure that it was a time shift. Why? My RPI crashed, I pulled the plug and restarted it. Domoticz started before the OS completed the NTP sync. So when the NTP sync completed the time shifted and this happened in Domoticz. I have fixed my setup by preventing Domoticz from starting while a NTP sync is still in progress. But there a many other reasons why a time shift can happen and that's why they call the current implementation a bad practice ;-)

**Comment by gizmocuz** [comment]
The timeshift should happen at 02:00am ?

**Comment by sander1988** [comment]
> The timeshift should happen at 02:00am ?

Can you explain why you think it's around 02:00AM? Maybe I can provide you some more details.

The time shift is here (copied from my previous message):
```
2021-03-17 08:17:15.296  Status: RFLink: Using serial port: /dev/ttyACM0
2021-03-17 08:32:18.949  Status: MQTT: Connecting to 127.0.0.1:1883
```

Which matches my syslog:
```
Mar 17 08:32:18 domoticz systemd-timesyncd[320]: Synchronized to time server for the first time 1.2.3.4:123 (private server removed).
```

**Comment by gizmocuz** [comment]
@sander1988 , I don't know from which country you are, but in the Netherlands we change the local time offset twice a year
And that should the only big time offset change be
In your case, it could be not related to this time change, and by the look of it, it could very well be a mainworker that has been blocked by a hardware/plugin

**Comment by sander1988** [comment]
I live in the Netherlands; so the same DST settings are used in my setup.

> In your case, it could be not related to this time change, and by the look of it, it could very well be a mainworker that has been blocked by a hardware/plugin

No, this issue can happen for 3 reasons (which I have seen so far)...

1. DST change twice a year (but that's not the cause this time).
2. Booting the server, in this case a RPI which has no hardware clock, so there will be a time shift a few seconds after boot up (when the NTP server performs it's initial sync). **That's what happened here.**
3. The user (like me) testing some other issue (not Domoticz related) which involves changing the the OS clock.

**Comment by gizmocuz** [comment]
Strange, in the startup configuration /etc/init.d/domoticz.sh it says

# Required-Start:    $network $remote_fs $syslog $time

**Comment by waltervl** [comment]
This time shift causing a crash (especially after a cold boot) is a known issue (see wiki https://www.domoticz.com/wiki/Troubleshooting#Domoticz_stops_after_cold_boot for some workarounds).

**Comment by kiddigital** [comment]
Also see issue #4700 , related to above raised point 1; DST time change twice a year due to Daylight savings change.

**Comment by gizmocuz** [comment]
@waltervl , Don't experience issues with DST changes here... should work correctly

**Comment by sander1988** [comment]
Regarding the original issue...

I also haven't seen DST issue anymore (for atleast the last 2 DST changes).

The boot issue fixed / worked around by setting the right systemd service dependencies, mine looks like this now:
```
[Unit]
Description=Domoticz Home Automation
After=network.target time-sync.target

[Service]
...
```

But the issue itself isn't fixed, but that's due to the current implementation (clock() should be used). It still see crashes on time shifts. I see this happening in 2 scenarios but not really a problem as it's in a test environment:
- Change the system clock (for testing other things in my case).
- VM operations (I have several VMs running with Domoticz).


So no priority for me and a won't fix is also OK.

**Comment by waltervl** [comment]
I also do not see DST issues. I only see issues on the forum were people have troubles starting Domoticz after a cold boot (eg power failure) that took a long time (> 1 hour?) after shutdown.

**Comment by gizmocuz** [comment]
@waltervl , that is strange as well... They should check if there NTP client is working or use a system-d startup file like above

**Comment by waltervl** [comment]
But a lot of them do not know they have to do that...
I also see that a lot of them work with \etc\init.d\domoticz.sh were this also is not covered yet without modification of that file.

**Comment by gizmocuz** [comment]
I always grap a raspberry pi, put raspbian on it, configure timezone, install domoticz (via sudo or docker) and that is it
No issues
Maybe they use wifi instead of a direct ethernet connection

I aggree , best it to use the systemd startup method, there was some discussion before and I asked to make a PR, but in the end nobody did this
